velox/type/Variant.cpp

```diff
@@ -583,12 +584,43 @@ uint64_t variant::hash() const {
   if (a.isNull() or b.isNull()) {
     return false;
   }
+
+  // Use relative epsilon comparison.
+  // See
+  // https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/
+  // for details.
+
   if (a.kind_ == TypeKind::REAL) {
-    return fabs(a.value<TypeKind::REAL>() - b.value<TypeKind::REAL>()) <
-        kEpsilon;
+    float f1 = a.value<TypeKind::REAL>();
+    float f2 = b.value<TypeKind::REAL>();
+
+    // Check if the numbers are really close -- needed
+    // when comparing numbers near zero.
+    if (fabs(f1 - f2) < kEpsilon) {
+      return true;
+    }
+
+    float a1 = abs(f1);
+    float a2 = abs(f2);
+    float largest = a1 > a2 ? a1 : a2;
+
+    return fabs(f1 - f2) <= largest * FLT_EPSILON;
+  } else {
+    double d1 = a.value<TypeKind::DOUBLE>();
+    double d2 = b.value<TypeKind::DOUBLE>();
+
+    // Check if the numbers are really close -- needed
+    // when comparing numbers near zero.
+    if (fabs(d1 - d2) < kEpsilon) {
+      return true;
+    }
+
+    double a1 = abs(d1);
+    double a2 = abs(d2);
+    double largest = a1 > a2 ? a1 : a2;
+
+    return fabs(d1 - d2) <= largest * FLT_EPSILON;
```

Comment:
could also factor this out to a tiny templated helper method to avoid duplication